chemreg/compound/models.py

```diff
@@ -48,3 +51,121 @@ class DefinedCompound(BaseCompound):
+    Label = String (Less than 100 characters, unique, required field)
+    Short Description = String (Less than 500 characters, required field)
+    Long Description = TEXT (required field)
+    """
+
+    name = models.SlugField(
+        max_length=49,
+        verbose_name="name",
+        help_text="Query structure type name",
+        unique=True,
+        blank=False,
+    )
+    label = models.CharField(
+        max_length=99,
+        verbose_name="label",
+        help_text="Query structure type label",
+        unique=True,
+        blank=False,
+    )
+    short_description = models.CharField(
+        max_length=499,
+        verbose_name="short description",
+        help_text="Query structure type short description",
+        blank=False,
+    )
+    long_description = models.TextField(
+        verbose_name="long description",
+        help_text="Query structure type long description",
+        blank=False,
+    )
+
+    def __str__(self):
+        return self.label
+
+    # from chemreg.compound.models import QueryStructureType
+    # qst=QueryStructureType(name="$ slug",label="a label",short_description="something short",long_description="much longer" * 20)
+
+    def save(self, *args, **kwargs):
+        unique_slugify(self, self.name, "name")
+        super(QueryStructureType, self).save()
+
+
+def unique_slugify(
+    instance, value, slug_field_name="slug", queryset=None, slug_separator="-"
+):
+    """
+    Calculates and stores a unique slug of ``value`` for an instance.
+
+    ``slug_field_name`` should be a string matching the name of the field to
+    store the slug in (and the field to check against for uniqueness).
+
+    ``queryset`` usually doesn't need to be explicitly provided - it'll default
+    to using the ``.all()`` queryset from the model's default manager.
+    """
+    slug_field = instance._meta.get_field(slug_field_name)
+
+    slug = getattr(instance, slug_field.attname)
+    slug_len = slug_field.max_length
+
+    # Sort out the initial slug, limiting its length if necessary.
+    slug = slugify(value)
+    if slug_len:
+        slug = slug[:slug_len]
+    slug = _slug_strip(slug, slug_separator)
+    original_slug = slug
+
+    # Create the queryset if one wasn't explicitly provided and exclude the
+    # current instance from the queryset.
+    if queryset is None:
+        queryset = instance.__class__._default_manager.all()
+    if instance.pk:
+        queryset = queryset.exclude(pk=instance.pk)
+
+    # Find a unique slug. If one matches, at '-2' to the end and try again
+    # (then '-3', etc).
+    next = 2
+    while not slug or queryset.filter(**{slug_field_name: slug}):
+        slug = original_slug
+        end = "%s%s" % (slug_separator, next)
+        if slug_len and len(slug) + len(end) > slug_len:
+            slug = slug[: slug_len - len(end)]
+            slug = _slug_strip(slug, slug_separator)
+        slug = "%s%s" % (slug, end)
+        next += 1
```

Comment:
Automatically generating the slug like this is deceptively challenging. This is subject to race conditions where two slugs could be made at the same time with the same count. In doing so, we have an unbounded loop to call to the database. The performance of the application goes from O(1) for the uniqueness check to a worst case scenario of O(n) where n is the number of repeated slugs.

I'm not sure all this is worth it. If we need to generate slugs on-the-fly in the future, I think we will need a more atomic operation to do so. For now, the ticket does not want this automated creation, so I think we can take this complexity out.